packages/gatsby/src/redux/reducers/queries.ts

```diff
@@ -0,0 +1,209 @@
+      // TODO: use hash instead of a query text
+      const { componentPath, query } = action.payload
+      const component = registerComponent(state, componentPath)
+      if (component.query !== query) {
+        // Invalidate all pages associated with a component when query text changes
+        component.pages.forEach(queryId => {
+          const query = state.trackedQueries.get(queryId)
+          if (query) {
+            query.dirty = setFlag(query.dirty, FLAG_DIRTY_TEXT)
+          }
+        })
+        component.query = query
+      }
+      return state
+    }
+    case `QUERY_EXTRACTION_GRAPHQL_ERROR`:
+    case `QUERY_EXTRACTION_BABEL_ERROR`:
+    case `QUERY_EXTRACTION_BABEL_SUCCESS`: {
+      const { componentPath } = action.payload
+      const component = registerComponent(state, componentPath)
+      const set = action.type !== `QUERY_EXTRACTION_BABEL_SUCCESS`
+      component.errors = setFlag(component.errors, FLAG_ERROR_BABEL, set)
+      return state
+    }
+    case `REPLACE_STATIC_QUERY`: {
+      // Only called when static query text has changed, so no need to compare
+      // TODO: unify the behavior?
+      const query = registerQuery(state, action.payload.id)
+      query.dirty = setFlag(query.dirty, FLAG_DIRTY_TEXT)
+      return state
+    }
+    case `REMOVE_STATIC_QUERY`: {
+      state.trackedQueries.delete(action.payload)
+      return state
+    }
+    case `CREATE_COMPONENT_DEPENDENCY`: {
+      const { path: queryId, nodeId, connection } = action.payload
+      if (nodeId) {
+        const queryIds = state.byNode.get(nodeId) ?? new Set<QueryId>()
+        queryIds.add(queryId)
+        state.byNode.set(nodeId, queryIds)
+      }
+      if (connection) {
+        const queryIds =
+          state.byConnection.get(connection) ?? new Set<QueryId>()
+        queryIds.add(queryId)
+        state.byConnection.set(connection, queryIds)
+      }
+      return state
+    }
+    case `QUERY_START`: {
+      // Reset data dependencies as they will be updated when running the query
+      const { path } = action.payload
+      state.byNode.forEach(queryIds => {
+        queryIds.delete(path)
+      })
+      state.byConnection.forEach(queryIds => {
+        queryIds.delete(path)
+      })
+      return state
+    }
+    case `CREATE_NODE`:
+    case `DELETE_NODE`: {
+      const node = action.payload
+      const queriesByNode = state.byNode.get(node.id) ?? []
+      const queriesByConnection =
+        state.byConnection.get(node.internal.type) ?? []
+
+      queriesByNode.forEach(queryId => {
+        const query = state.trackedQueries.get(queryId)
+        if (query) {
+          query.dirty = setFlag(query.dirty, FLAG_DIRTY_DATA)
+        }
+      })
+      queriesByConnection.forEach(queryId => {
+        const query = state.trackedQueries.get(queryId)
+        if (query) {
+          query.dirty = setFlag(query.dirty, FLAG_DIRTY_DATA)
+        }
+      })
+      return state
+    }
```

Comment:
Marking both - static and page queries as dirty when nodes queried during the previous query execution change. Query <-> Node dependencies are registered above via `CREATE_COMPONENT_DEPENDENCY` action (that's the same logic as in the old `component-data-dependencies`, just combined with other actions in this reducer)

During bootstrap, this is a no-op as we have a clean cache and no data dependencies.